main/refer.c

```diff
@@ -0,0 +1,527 @@
+AST_VECTOR(, const struct ast_refer_tech *) refer_techs;
+
+static int refer_data_cmp_fn(void *obj, void *arg, int flags)
+{
+	const struct refer_data *one = obj, *two = arg;
+	if (one == NULL || one->name == NULL || two == NULL || two->name == NULL) {
+		return 0;
+	}
+	return !strcasecmp(one->name, two->name) ? CMP_MATCH | CMP_STOP : 0;
+}
+
+static void refer_data_destructor(void *obj)
+{
+	struct refer_data *data = obj;
+	ast_string_field_free_memory(data);
+}
+
+static void refer_destructor(void *obj)
+{
+	struct ast_refer *refer = obj;
+
+	ast_string_field_free_memory(refer);
+	ao2_cleanup(refer->vars);
+}
+
+struct ast_refer *ast_refer_alloc(void)
+{
+	struct ast_refer *refer;
+
+	if (!(refer = ao2_alloc(sizeof(*refer), refer_destructor))) {
+		return NULL;
+	}
+
+	if (ast_string_field_init(refer, 128)) {
+		ao2_ref(refer, -1);
+		return NULL;
+	}
+
+	refer->vars = ao2_container_alloc_list(AO2_ALLOC_OPT_LOCK_MUTEX, 0,
+		NULL, refer_data_cmp_fn);
+	if (!refer->vars) {
+		ao2_ref(refer, -1);
+		return NULL;
+	}
+	refer->to_self = 0;
+
+	return refer;
+}
+
+struct ast_refer *ast_refer_ref(struct ast_refer *refer)
+{
+	ao2_ref(refer, 1);
+	return refer;
+}
+
+struct ast_refer *ast_refer_destroy(struct ast_refer *refer)
+{
+	ao2_ref(refer, -1);
+	return NULL;
+}
+
+int ast_refer_set_to(struct ast_refer *refer, const char *fmt, ...)
+{
+	va_list ap;
+
+	va_start(ap, fmt);
+	ast_string_field_build_va(refer, to, fmt, ap);
+	va_end(ap);
+
+	return 0;
+}
+
+int ast_refer_set_from(struct ast_refer *refer, const char *fmt, ...)
+{
+	va_list ap;
+
+	va_start(ap, fmt);
+	ast_string_field_build_va(refer, from, fmt, ap);
+	va_end(ap);
+
+	return 0;
+}
+
+int ast_refer_set_refer_to(struct ast_refer *refer, const char *fmt, ...)
+{
+	va_list ap;
+
+	va_start(ap, fmt);
+	ast_string_field_build_va(refer, refer_to, fmt, ap);
+	va_end(ap);
+
+	return 0;
+}
+
+int ast_refer_set_to_self(struct ast_refer *refer, int val)
+{
+	refer->to_self = val;
+	return 0;
+}
+
+int ast_refer_set_tech(struct ast_refer *refer, const char *fmt, ...)
+{
+	va_list ap;
+
+	va_start(ap, fmt);
+	ast_string_field_build_va(refer, tech, fmt, ap);
+	va_end(ap);
+
+	return 0;
+}
+
+int ast_refer_set_endpoint(struct ast_refer *refer, const char *fmt, ...)
+{
+	va_list ap;
+
+	va_start(ap, fmt);
+	ast_string_field_build_va(refer, endpoint, fmt, ap);
+	va_end(ap);
+
+	return 0;
+}
+
+const char *ast_refer_get_refer_to(const struct ast_refer *refer)
+{
+	return refer->refer_to;
+}
+
+const char *ast_refer_get_from(const struct ast_refer *refer)
+{
+	return refer->from;
+}
+
+const char *ast_refer_get_to(const struct ast_refer *refer)
+{
+	return refer->to;
+}
+
+int ast_refer_get_to_self(const struct ast_refer *refer)
+{
+	return refer->to_self;
+}
+
+const char *ast_refer_get_tech(const struct ast_refer *refer)
+{
+	return refer->tech;
+}
+
+const char *ast_refer_get_endpoint(const struct ast_refer *refer)
+{
+	return refer->endpoint;
+}
+
+static struct refer_data *refer_data_alloc(void)
```

Comment:
Per my earlier comment, I think this should take in the name and value and allocate the specific size needed.

Reply:
Done.